src/System.Net.Http/src/System/Net/Http/Managed/AuthenticationHelper.Digest.cs

```diff
@@ -0,0 +1,374 @@
+                if (index1 != -1)
+                {
+                    // Get index of auth if present in qop string
+                    int index2 = digestResponse.Parameters[Qop].IndexOf(Auth);
+
+                    // If index2 < index1, auth option is available
+                    // If index2 == index1, check if auth option available later in string after auth-int.
+                    if (index2 == index1)
+                    {
+                        index2 = digestResponse.Parameters[Qop].IndexOf(Auth, index1 + AuthInt.Length);
+                        if (index2 == -1)
+                        {
+                            qop = AuthInt;
+                        }
+                    }
+                }
+            }
+
+            // Set cnonce
+            string cnonce = GetRandomAlphaNumericString();
+
+            // Calculate response
+            string a1 = credential.UserName + Colon + realm + Colon + credential.Password;
+            if (digestResponse.Parameters[Algorithm].Contains(Sha256Sess))
+            {
+                a1 = ComputeHash(a1, algorithm) + Colon + digestResponse.Parameters[Nonce] + Colon + cnonce;
+            }
+
+            string a2 = request.Method.Method + Colon + request.RequestUri.PathAndQuery;
+            if (qop == AuthInt)
+            {
+                a2 = a2 + Colon + ComputeHash(request.Content.ReadAsStringAsync().Result, algorithm);
+            }
+
+            string response = ComputeHash(ComputeHash(a1, algorithm) + Colon +
+                                        digestResponse.Parameters[Nonce] + Colon +
+                                        digestResponse.NonceCount.ToString("x8") + Colon +
+                                        cnonce + Colon +
+                                        qop + Colon +
+                                        ComputeHash(a2, algorithm), algorithm);
+
+            // Add response
+            sb.AppendKeyValue(Response, response);
+
+            // Add algorithm
+            sb.AppendKeyValue(Algorithm, algorithm, includeQuotes: false);
+
+            // Add opaque
+            sb.AppendKeyValue(Opaque, digestResponse.Parameters[Opaque]);
+
+            // Add qop
+            sb.AppendKeyValue(Qop, qop, includeQuotes: false);
+
+            // Add nc
+            sb.AppendKeyValue(NC, digestResponse.NonceCount.ToString("x8"), includeQuotes: false);
+
+            // Add cnonce
+            sb.AppendKeyValue(CNonce, cnonce, includeComma: false);
+
+            return StringBuilderCache.GetStringAndRelease(sb);
+        }
+
+        private static string GetRandomAlphaNumericString()
+        {
+            using (RNGCryptoServiceProvider rng = new RNGCryptoServiceProvider())
+            {
+                const int length = 16;
+                StringBuilder sb = StringBuilderCache.Acquire(length);
+                for (int i = 0; i < length; i++)
+                {
+                    byte[] randomNumber = new byte[1];
```

Comment:
If this switches to Random, we can use the Next(Span) overload once it's been implemented.  Otherwise, hopefully we can add https://github.com/dotnet/corefx/issues/22690 and then switch to using that.